feat:  Matrix generator where a generator can reference items of another one

KojoRising opened the pull request:
Link - https://github.com/argoproj/applicationset/issues/530

Fixes [ISSUE #530]
Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

rishabh625 commented:
it looks good to me, plz include test case, also it would be great to have the usage documented with example as only second generator refers params of 1st

KojoRising commented:
@rishabh625  - Awesome sounds great!. Writing up the tests right now. Looking through the ArgoCD Repo as to where I can update edit the Matrix generator docs, but I don't see them under "applicationset" directory, and the regular docs don't seem to mention them. So my question is: 

- Are the ApplicationSet docs still hosted from the ApplicationSet repository? 
- If so, should I create a separate PR to add documentation there?

Thanks, and let me know!

rishabh625 commented:
I guess you need to update docs here, as docs is moved https://github.com/argoproj/argo-cd/pull/8955

Are we going to change hosting of application set docs 
May be @jgwest @jannfis @crenshaw-dev  can answer.

KojoRising commented:
Thanks @crenshaw-dev, found it! 

I've updated the Matrix Generator documentation to include the following: 

1) An example which shows one child generator consuming parameters from another child generator
2) Restrictions on the parameter consumption. In particular:
- The child generator which consumes parameters **must come after** the child generator that produces the parameters
- Both child generators' cannot consume parameters from one-another (ie. circular dependency issue).

I added the restrictions to the bottom (where the other Matrix Generator Restrictions are), but let me know if I should explicitly make those restrictions clear in the section with the example.

Thanks!

Lobstrosity commented on applicationset/generators/generator_spec_processor.go:
```diff
@@ -81,3 +91,33 @@ func mergeGeneratorTemplate(g Generator, requestedGenerator *argoprojiov1alpha1.
+		log.WithError(err).WithField("requestedGenerator", requestedGenerator).Error("error marshalling requested generator for interpolation")
+		return err
+	}
+	tmpParams := make(map[string]string)
+	for _, currParam := range params {
+		for k, v := range currParam {
+			tmpParams[k] = v
+		}
+	}
```
After this block of code, won’t `tmpParams` contain whatever was in the last map in the `params` list for a given key? Meaning that the templating below would always interpolate the last value for a given key instead of _each_ value for a given key?

If I’m understanding the use case, it seems `params` should be a map (instead of a list of maps) and then the matrix generator code would look something like:

```go
g0, err := m.getParams(appSetGenerator.Matrix.Generators[0], appSet, map[string]string{})

for _, a := range g0 {
  g1, err := m.getParams(appSetGenerator.Matrix.Generators[1], appSet, a)

  for _, b := range g1 {
    val, err := utils.CombineStringMaps(a, b)
    if err != nil {
      return nil, err
    }
    res = append(res, val)
  }
}
```

…where `g1` is determined for each param map `a` from `g0`.

KojoRising replied:
So this is how the "params" function parameter value looks like 
<img width="1273" alt="image" src="https://user-images.githubusercontent.com/29113963/163661670-ebd236bf-cb95-41c0-8ce6-ad2f92f7d0c1.png">


tmpParams ends up like this: 
<img width="1328" alt="image" src="https://user-images.githubusercontent.com/29113963/163661762-48cf436f-0e1b-49af-aed4-d4615f4b9f63.png">

KojoRising replied:
Also to your point, where the interpolation would only pick up the last value, I also tried it w/ {{path[3]}}, and this is what I got: 

<img width="1021" alt="image" src="https://user-images.githubusercontent.com/29113963/163661896-b29a6fdc-ce9e-4936-b69c-9f6282b70a38.png">

<img width="1277" alt="image" src="https://user-images.githubusercontent.com/29113963/163661921-c9a12992-d5ab-4551-a678-ee2e2fec08b7.png">

Lobstrosity commented on applicationset/generators/generator_spec_processor_test.go:
```diff
@@ -0,0 +1,300 @@
+	}
+	argoCDServiceMock := argoCDServiceMock{mock: &mock.Mock{}}
+
+	argoCDServiceMock.mock.On("GetDirectories", mock.Anything, mock.Anything, mock.Anything).Return(cases[0].repoApps, cases[0].repoError)
```
You define several `cases` but only appear to use the first one.

KojoRising replied:
Ahh yeah, this was more for testing the `GetRelevantGenerators` function (not mine). I'm fairly new to Go (I'm more on the ArgoCD Yaml side :), so I was using some tests that were used for the Git Generators (where they were specifically testing the Git Generator).

My test cases were below in the `TestGetRelevantGenerators` function:

<img width="911" alt="image" src="https://user-images.githubusercontent.com/29113963/164109474-673c8a96-7665-4fda-8ee5-7899178dda5d.png">

Anyways, I cleaned up the above function you mentioned. Thanks for pointing it out!

Lobstrosity commented on applicationset/generators/matrix.go:
```diff
@@ -44,11 +44,11 @@ func (m *MatrixGenerator) GenerateParams(appSetGenerator *argoprojiov1alpha1.App
 		return nil, err
 	}
-	g1, err := m.getParams(appSetGenerator.Matrix.Generators[1], appSet)
+	g1, err := m.getParams(appSetGenerator.Matrix.Generators[1], appSet, g0)
```
Regarding your “I think it'll always be an array of size 1” comment from [the other thread](https://github.com/argoproj/argo-cd/pull/9080/files#r851585457), isn’t `g0` _all_ items from the first generator? And doesn’t it eventually become the `params` parameter to your new `interpolateGenerator` function? I’m unclear how the new example you added to `Generators-Matrix.md` can be satisfied unless `g1` is evaluated for each item in `g0`, instead of only for the last item in `g0` (which we seem to agree is what’s happening based on that loop in `interpolateGenerator`).

I suggest additional test case(s) in `matrix_test.go` to validate the complete results from a matrix generator using this functionality (specifically where the first generator yields multiple items).

I’m very interested in this PR for some of my own use cases and I’d try to help out but so far, I’ve been unsuccessful getting a local environment up and running.

KojoRising replied:
Hi @Lobstrosity,

So as to the example you were referring to, I'm not sure if you were able to check my *TestInterpolateGenerator* function. I had the following example which was very similar to the example I had in `Generators-Matrix.md` (actually it used more parameters than the example). As you can see in the below test case, the cluster-generator is using 3 different parameters that should be created by the git generator. 

<img width="896" alt="image" src="https://user-images.githubusercontent.com/29113963/164104082-5b9cf149-c756-4b32-9c84-cc8f7b1eb10c.png">

If `g1` does need to be evaluated for every item in `g0`, (by item, I'm assuming we're both referring to parameter generated for that particular generator), and that it currently only evaluates the last item, then wouldn't you agree that:
1) All the keys for this label-selector's values - Be evaluated with the last item in the params list (ie. `app3`)
2) All these test-cases be failing? 

<img width="935" alt="image" src="https://user-images.githubusercontent.com/29113963/164105635-f276a64c-e9b0-4664-8b32-e9f7dd99fcd1.png">

KojoRising replied:
Also though, I will try your above code snippet and see if I can observe a difference. 

```yaml
g0, err := m.getParams(appSetGenerator.Matrix.Generators[0], appSet, map[string]string{})

for _, a := range g0 {
  g1, err := m.getParams(appSetGenerator.Matrix.Generators[1], appSet, a)

  for _, b := range g1 {
    val, err := utils.CombineStringMaps(a, b)
    if err != nil {
      return nil, err
    }
    res = append(res, val)
  }
}
```

rishabh625 replied:
Also @KojoRising m unsure the example in Generators-Matrix.md is workable with this code. i would be good to see a e2e test.

Lobstrosity commented on applicationset/generators/generator_spec_processor_test.go:
```diff
@@ -0,0 +1,300 @@
+		},
+	}
+	//mockGitGenerator := getMockGitGenerator()
+	gitGeneratorParams := []map[string]string{
+		{"path": "p1/p2/app3", "path.basename": "app3", "path[0]": "p1", "path[1]": "p2", "path.basenameNormalized": "app3"},
+	}
```
@KojoRising Regarding [this comment](https://github.com/argoproj/argo-cd/pull/9080#discussion_r853506892) and the screenshot with the arrows:

This `gitGeneratorParams` mock represents a `git` generator that yields a single result. My concern is, what happens when the `git` generator yields multiple results? _All_ of those results would be in `g0` over in `matrix.go` but `interpolateGenerator` is only interpolating values from the first result.

KojoRising replied:
@Lobstrosity 

Ahh, got it got it. Yes, you’re right, for multiple paths specified in a git generator, it wouldnt pick those up. 

I’ll update the code/test-cases to support this. Thank you for pointing it out!

KojoRising commented:
@Lobstrosity, @rishabh625, @crenshaw-dev Need some input from your folks here. This issue is in line with what Lobstrosity pointed out above. So roughly

1) When I replace the templated string, I first get `kubernetes.io/app-type: argo-workflows` 
![image](https://user-images.githubusercontent.com/29113963/164293682-ae373fde-f19c-475b-8142-140632d30546.png)

![image](https://user-images.githubusercontent.com/29113963/164293467-ded05807-9a48-49a9-b087-c9bc6bf8ea3e.png)

2) The issue is that on the 2nd string replacement/interpolation, it searches for the same `{{path.basename}}` parameter and tries to replace it. But it's already replaced it with `argo-workflows`, so it doesn't get replaced. 


And even if I manage to save a copy of the string, the following is obviously going to be invalid YAML: 
![image](https://user-images.githubusercontent.com/29113963/164294094-b46b8323-fbb7-455d-ba8e-11457914b125.png)

==========================================================================================

So my question for y'all is do you think the below: 

![image](https://user-images.githubusercontent.com/29113963/164292592-081f48b9-0fda-4f55-84f4-755504b884f5.png)


Should resolve to this: 
![image](https://user-images.githubusercontent.com/29113963/164291304-ac6f9a7c-867e-431f-8d58-7d7e211f73a7.png)

I can't really think of another way of representing it?. Unless I create two appsets, or two matrix generators with that interpolation resolved.

rishabh625 commented:
I guess just representing  it as [set-based ](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#resources-that-support-set-based-requirements) will not solve the issue as issue is not just limited to matchLabels #2 child generator can refer parameters of #1 child generator anywhere, consider if git and list generators are combined, it will get more tricky,      


 I'll get back to u

KojoRising commented:
Also, I want to to propose two solutions:

1) Enforce having the cluster-generator 1st. In this case, we can easily add another path-object to the git-generator's paths array:
<img width="818" alt="image" src="https://user-images.githubusercontent.com/29113963/164314807-cd67c82e-98d6-4dab-92c3-0c6a9e7f1a48.png">

Resolving to this:
<img width="672" alt="image" src="https://user-images.githubusercontent.com/29113963/164314925-a2b68bce-58e2-495c-918c-41273aabcd0d.png">

Drawbacks: 
1) Not flexible enough for many generators
2) Enforcing this will have to be heavily documented


2) Instead of just {{path.basename}}, on a per-cluster basis we can access the clusters like an array index. So if we had two clusters, each labelled w/ "argo-workflows" & "prometheus-operator", we could use the appset like this:

<img width="707" alt="image" src="https://user-images.githubusercontent.com/29113963/164315409-c0b22f57-7fe0-4d1a-ad7c-118db07875dc.png">


Drawbacks: 
1) "path[n]" will probably end up being confusing, since it would be like "path[c][n]" (where c is the cluster, and n is a subpath in the full-path name)
2) If you wanted to specify a cluster for every {{path.basename}}, that would now be incredibly tedious.

Lobstrosity commented:
@KojoRising Is it not possible to re-evaluate the 2nd generator for each result from the 1st generator? (If I understood one of your previous comments, I think you were saying that the first interpolation pass essentially hard-codes the values into the generator so that there’s nothing to interpolate on the next pass?)

Maybe have `interpolateGenerator` return a new generator with the values interpolated into it, instead of modifying the generator passed to it?

Lobstrosity commented:
If the 2nd generator is re-evaluated for each result from 1st generator, I don’t think the spec would need to change at all and you wouldn’t need to enforce which type of generator was first (`clusters` vs. `git`).

Re-evaluating 2nd generator each time is where the most compelling functionality comes in, IMO. It basically allows for inner joins of the 2 generator’s data sources.

Example use case: https://gist.github.com/Lobstrosity/bd6369ec09adc2b9d1db08619c5193a3

rishabh625 commented on applicationset/generators/generator_spec_processor.go:
```diff
@@ -31,7 +33,17 @@ func Transform(requestedGenerator argoprojiov1alpha1.ApplicationSetGenerator, al
 			continue
 		}
-
+		if len(genParams) != 0 {
```
unless g.GenerateParams is called as in line no 47, genParams wouldn't be always zero?

KojoRising replied:
Not for the Matrix generator. It will be 0 when called from the applicationset_controller.go file, but when it's called from matrix.go I pass in the previous generator's params.

rishabh625 commented on applicationset/generators/generator_spec_processor.go:
```diff
@@ -81,3 +91,30 @@ func mergeGeneratorTemplate(g Generator, requestedGenerator *argoprojiov1alpha1.
+// "params" parameter is an array, where each index corresponds to a generator. Each index contains a map w/ that generator's parameters.
+// Since the Matrix generator currently only allows 2 child generators, effectively this params array will always be size 1 (containing the 1st child generator's params)
+// Uses similar process for interpreting values as the actual Application Template
+func interpolateGenerator(requestedGenerator *argoprojiov1alpha1.ApplicationSetGenerator, params []map[string]string) error {
```
@KojoRising like @Lobstrosity said returning generator from interpolation would solve ur problem of resusbstituting as original generator will not be replaced like this

```
func interpolateGenerator(requestedGenerator *argoprojiov1alpha1.ApplicationSetGenerator, params []map[string]string) (argoprojiov1alpha1.ApplicationSetGenerator, error) {
	newrequestedGenerator := *requestedGenerator
	tmplBytes, err := json.Marshal(newrequestedGenerator)
	if err != nil {
		log.WithError(err).WithField("requestedGenerator", requestedGenerator).Error("error marshalling requested generator for interpolation")
		return newrequestedGenerator, err
	}

	render := utils.Render{}
	fstTmpl := fasttemplate.New(string(tmplBytes), "{{", "}}")
	replacedTmplStr, err := render.Replace(fstTmpl, params[0], true)
	if err != nil {
		log.WithError(err).WithField("interpolatedGeneratorString", replacedTmplStr).Error("error interpolating generator with other generator's parameter")
		return newrequestedGenerator, err
	}

	err = json.Unmarshal([]byte(replacedTmplStr), requestedGenerator)
	if err != nil {
		log.WithError(err).WithField("requestedGenerator", requestedGenerator).Error("error unmarshalling requested generator for interpolation")
		return newrequestedGenerator, err
	}
	return newrequestedGenerator, nil
}
```

rishabh625 commented on applicationset/generators/generator_spec_processor.go:
```diff
@@ -31,7 +33,17 @@ func Transform(requestedGenerator argoprojiov1alpha1.ApplicationSetGenerator, al
+				continue
+			}
+		}
 		params, err := g.GenerateParams(&requestedGenerator, appSet)
```
in addition to above change replace requestedGenerator with newrequestedgenerator

KojoRising commented:
Ok, thanks @Lobstrosity @rishabh625. This is working now. For context, this is the example appset I used:

````yaml
apiVersion: argoproj.io/v1alpha1
kind: ApplicationSet
metadata:
  name: cluster-git
spec:
  generators:
    - matrix:
        generators:
          - clusters:
              selector:
                matchLabels:
                  argocd.argoproj.io/secret-type: cluster
                  kubernetes.io/app-type: '{{path.basename}}'
          - git:
              repoURL: https://github.com/argoproj/argo-cd.git
              revision: HEAD
              directories:
                - path: 'applicationset/examples/matrix/cluster-addons/argo-workflows'
                - path: 'applicationset/examples/matrix/cluster-addons/prometheus-operator'
````

With the following secrets: 
![Pasted Graphic](https://user-images.githubusercontent.com/29113963/164765268-84257910-1772-407d-abee-fa3e97402a92.png)


Getting the following Applications (sorry, can't shown cluster-names, cause corporate :)):
![image](https://user-images.githubusercontent.com/29113963/164765432-e7a25693-5ac2-4a8e-bbb3-9db51b13749a.png)

KojoRising commented:
Gitea/SCM tests are failing the unit tests. Waiting for this PR (https://github.com/argoproj/argo-cd/pull/9194) to be merged in to hopefully restore intended functionality to those test cases.

Lobstrosity commented:
@KojoRising In [this comment](https://github.com/argoproj/argo-cd/pull/9080#issuecomment-1106722768), the first generator has the parameterization. I thought only the second generator supports it?

KojoRising commented:
@Lobstrosity Ahh, yeah, you're right. I was trying some other stuff out in the yaml, and must have re-ordered it wrong when pasting it up. Here's how it looks intra-cluster: 

<img width="709" alt="image" src="https://user-images.githubusercontent.com/29113963/165176549-e71fa431-a10f-4600-b453-86811263e6bc.png">

jannis-a commented:
Great work so far, kudos! What's still missing for this contribution, anything I could possibly help with? :)

rishabh625 commented:
@KojoRising need to merge master here

KojoRising commented:
Apologies, was out since the middle of last week. 

Unfortunately, been running into the following issue with one of the Matrix Generator's test cases. It's related to the following test:

`returns error if there is more than one inner generator in the second base generator`

Been diagnosing it for about a day & change, and am still stuck at the moment (I am unfortunately not great at Go nor Go testing. Just a yaml engineer now :)).

What may be of note is the test `returns error if there is more than one inner generator in the first base generator` works without problem. This seems to be in line with the behaviour of the interpolatedGenerator, but unable to find a fix at the moment. 

@jannis-a If you get a sec to take a look, that'd be great! (not tryna pass off tests, but this is where I'm stuck at at the moment :))). But I'll continue to work on it through the week.

@rishabh625, @Lobstrosity if y'all have some input on this as well that'd be greatly appreciated. Thanks! 

```
/usr/local/go/bin/go test -c -o /private/var/folders/1t/f8k8j3c54vd77zvr_6ygxm3h0000gp/T/___1matrix_test_go.test github.com/argoproj/argo-cd/v2/applicationset/generators #gosetup
/usr/local/go/bin/go tool test2json -t /private/var/folders/1t/f8k8j3c54vd77zvr_6ygxm3h0000gp/T/___1matrix_test_go.test -test.v -test.paniconexit0 -test.run ^\QTestMatrixGenerate\E|\QTestMatrixGetRequeueAfter\E$
=== RUN   TestMatrixGenerate
=== RUN   TestMatrixGenerate/returns_error_if_there_is_more_than_one_inner_generator_in_the_second_base_generator
--- FAIL: TestMatrixGenerate (0.00s)
    --- FAIL: TestMatrixGenerate/returns_error_if_there_is_more_than_one_inner_generator_in_the_second_base_generator (0.00s)
panic: 

mock: Unexpected Method Call
-----------------------------

GenerateParams(*v1alpha1.ApplicationSetGenerator,*v1alpha1.ApplicationSet)
		0: &v1alpha1.ApplicationSetGenerator{List:(*v1alpha1.ListGenerator)(0xc000148b00), Clusters:(*v1alpha1.ClusterGenerator)(nil), Git:(*v1alpha1.GitGenerator)(0xc000703860), SCMProvider:(*v1alpha1.SCMProviderGenerator)(nil), ClusterDecisionResource:(*v1alpha1.DuckTypeGenerator)(nil), PullRequest:(*v1alpha1.PullRequestGenerator)(nil), Matrix:(*v1alpha1.MatrixGenerator)(nil), Merge:(*v1alpha1.MergeGenerator)(nil)}
		1: &v1alpha1.ApplicationSet{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"", GenerateName:"", Namespace:"", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), DeletionTimestamp:<nil>, DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ClusterName:"", ManagedFields:[]v1.ManagedFieldsEntry(nil)}, Spec:v1alpha1.ApplicationSetSpec{Generators:[]v1alpha1.ApplicationSetGenerator(nil), Template:v1alpha1.ApplicationSetTemplate{ApplicationSetTemplateMeta:v1alpha1.ApplicationSetTemplateMeta{Name:"", Namespace:"", Labels:map[string]string(nil), Annotations:map[string]string(nil), Finalizers:[]string(nil)}, Spec:v1alpha1.ApplicationSpec{Source:v1alpha1.ApplicationSource{RepoURL:"", Path:"", TargetRevision:"", Helm:(*v1alpha1.ApplicationSourceHelm)(nil), Kustomize:(*v1alpha1.ApplicationSourceKustomize)(nil), Directory:(*v1alpha1.ApplicationSourceDirectory)(nil), Plugin:(*v1alpha1.ApplicationSourcePlugin)(nil), Chart:""}, Destination:v1alpha1.ApplicationDestination{Server:"", Namespace:"", Name:"", isServerInferred:false}, Project:"", SyncPolicy:(*v1alpha1.SyncPolicy)(nil), IgnoreDifferences:[]v1alpha1.ResourceIgnoreDifferences(nil), Info:[]v1alpha1.Info(nil), RevisionHistoryLimit:(*int64)(nil)}}, SyncPolicy:(*v1alpha1.ApplicationSetSyncPolicy)(nil)}, Status:v1alpha1.ApplicationSetStatus{Conditions:[]v1alpha1.ApplicationSetCondition(nil)}}

The closest call I have is: 

GenerateParams(*v1alpha1.ApplicationSetGenerator,*v1alpha1.ApplicationSet)
		0: &v1alpha1.ApplicationSetGenerator{List:(*v1alpha1.ListGenerator)(nil), Clusters:(*v1alpha1.ClusterGenerator)(nil), Git:(*v1alpha1.GitGenerator)(0xc000703040), SCMProvider:(*v1alpha1.SCMProviderGenerator)(nil), ClusterDecisionResource:(*v1alpha1.DuckTypeGenerator)(nil), PullRequest:(*v1alpha1.PullRequestGenerator)(nil), Matrix:(*v1alpha1.MatrixGenerator)(nil), Merge:(*v1alpha1.MergeGenerator)(nil)}
		1: &v1alpha1.ApplicationSet{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"", GenerateName:"", Namespace:"", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), DeletionTimestamp:<nil>, DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ClusterName:"", ManagedFields:[]v1.ManagedFieldsEntry(nil)}, Spec:v1alpha1.ApplicationSetSpec{Generators:[]v1alpha1.ApplicationSetGenerator(nil), Template:v1alpha1.ApplicationSetTemplate{ApplicationSetTemplateMeta:v1alpha1.ApplicationSetTemplateMeta{Name:"", Namespace:"", Labels:map[string]string(nil), Annotations:map[string]string(nil), Finalizers:[]string(nil)}, Spec:v1alpha1.ApplicationSpec{Source:v1alpha1.ApplicationSource{RepoURL:"", Path:"", TargetRevision:"", Helm:(*v1alpha1.ApplicationSourceHelm)(nil), Kustomize:(*v1alpha1.ApplicationSourceKustomize)(nil), Directory:(*v1alpha1.ApplicationSourceDirectory)(nil), Plugin:(*v1alpha1.ApplicationSourcePlugin)(nil), Chart:""}, Destination:v1alpha1.ApplicationDestination{Server:"", Namespace:"", Name:"", isServerInferred:false}, Project:"", SyncPolicy:(*v1alpha1.SyncPolicy)(nil), IgnoreDifferences:[]v1alpha1.ResourceIgnoreDifferences(nil), Info:[]v1alpha1.Info(nil), RevisionHistoryLimit:(*int64)(nil)}}, SyncPolicy:(*v1alpha1.ApplicationSetSyncPolicy)(nil)}, Status:v1alpha1.ApplicationSetStatus{Conditions:[]v1alpha1.ApplicationSetCondition(nil)}}

Difference found in argument 0:

--- Expected
+++ Actual
@@ -1,3 +1,17 @@
 (*v1alpha1.ApplicationSetGenerator)({
- List: (*v1alpha1.ListGenerator)(<nil>),
+ List: (*v1alpha1.ListGenerator)({
+  Elements: ([]v1.JSON) (len=1) {
+   (v1.JSON) &JSON{Raw:*[123 34 99 108 117 115 116 101 114 34 58 34 67 108 117 115 116 101 114 34 44 34 117 114 108 34 58 34 85 114 108 34 125],}
+  },
+  Template: (v1alpha1.ApplicationSetTemplate) {
+   ApplicationSetTemplateMeta: (v1alpha1.ApplicationSetTemplateMeta) {
+    Name: (string) "",
+    Namespace: (string) "",
+    Labels: (map[string]string) <nil>,
+    Annotations: (map[string]string) <nil>,
+    Finalizers: ([]string) <nil>
+   },
+   Spec: (v1alpha1.ApplicationSpec) &ApplicationSpec{Source:ApplicationSource{RepoURL:,Path:,TargetRevision:,Helm:nil,Kustomize:nil,Directory:nil,Plugin:nil,Chart:,},Destination:ApplicationDestination{Server:,Namespace:,Name:,},Project:,SyncPolicy:nil,IgnoreDifferences:[]ResourceIgnoreDifferences{},Info:[]Info{},RevisionHistoryLimit:nil,}
+  }
+ }),
  Clusters: (*v1alpha1.ClusterGenerator)(<nil>),

Diff: 0: FAIL:  (*v1alpha1.ApplicationSetGenerator=&{0xc000148b00 <nil> 0xc000703860 <nil> <nil> <nil> <nil> <nil>}) != (*v1alpha1.ApplicationSetGenerator=&{<nil> <nil> 0xc000703040 <nil> <nil> <nil> <nil> <nil>})
	1: PASS:  (*v1alpha1.ApplicationSet=&{{ } {      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []} {[] {{  map[] map[] []} {{   nil nil nil nil } {   false}  nil [] [] <nil>}} <nil>} {[]}}) == (*v1alpha1.ApplicationSet=&{{ } {      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []} {[] {{  map[] map[] []} {{   nil nil nil nil } {   false}  nil [] [] <nil>}} <nil>} {[]}}) [recovered]
	panic: 

mock: Unexpected Method Call
-----------------------------

GenerateParams(*v1alpha1.ApplicationSetGenerator,*v1alpha1.ApplicationSet)
		0: &v1alpha1.ApplicationSetGenerator{List:(*v1alpha1.ListGenerator)(0xc000148b00), Clusters:(*v1alpha1.ClusterGenerator)(nil), Git:(*v1alpha1.GitGenerator)(0xc000703860), SCMProvider:(*v1alpha1.SCMProviderGenerator)(nil), ClusterDecisionResource:(*v1alpha1.DuckTypeGenerator)(nil), PullRequest:(*v1alpha1.PullRequestGenerator)(nil), Matrix:(*v1alpha1.MatrixGenerator)(nil), Merge:(*v1alpha1.MergeGenerator)(nil)}
		1: &v1alpha1.ApplicationSet{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"", GenerateName:"", Namespace:"", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), DeletionTimestamp:<nil>, DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ClusterName:"", ManagedFields:[]v1.ManagedFieldsEntry(nil)}, Spec:v1alpha1.ApplicationSetSpec{Generators:[]v1alpha1.ApplicationSetGenerator(nil), Template:v1alpha1.ApplicationSetTemplate{ApplicationSetTemplateMeta:v1alpha1.ApplicationSetTemplateMeta{Name:"", Namespace:"", Labels:map[string]string(nil), Annotations:map[string]string(nil), Finalizers:[]string(nil)}, Spec:v1alpha1.ApplicationSpec{Source:v1alpha1.ApplicationSource{RepoURL:"", Path:"", TargetRevision:"", Helm:(*v1alpha1.ApplicationSourceHelm)(nil), Kustomize:(*v1alpha1.ApplicationSourceKustomize)(nil), Directory:(*v1alpha1.ApplicationSourceDirectory)(nil), Plugin:(*v1alpha1.ApplicationSourcePlugin)(nil), Chart:""}, Destination:v1alpha1.ApplicationDestination{Server:"", Namespace:"", Name:"", isServerInferred:false}, Project:"", SyncPolicy:(*v1alpha1.SyncPolicy)(nil), IgnoreDifferences:[]v1alpha1.ResourceIgnoreDifferences(nil), Info:[]v1alpha1.Info(nil), RevisionHistoryLimit:(*int64)(nil)}}, SyncPolicy:(*v1alpha1.ApplicationSetSyncPolicy)(nil)}, Status:v1alpha1.ApplicationSetStatus{Conditions:[]v1alpha1.ApplicationSetCondition(nil)}}

The closest call I have is: 

GenerateParams(*v1alpha1.ApplicationSetGenerator,*v1alpha1.ApplicationSet)
		0: &v1alpha1.ApplicationSetGenerator{List:(*v1alpha1.ListGenerator)(nil), Clusters:(*v1alpha1.ClusterGenerator)(nil), Git:(*v1alpha1.GitGenerator)(0xc000703040), SCMProvider:(*v1alpha1.SCMProviderGenerator)(nil), ClusterDecisionResource:(*v1alpha1.DuckTypeGenerator)(nil), PullRequest:(*v1alpha1.PullRequestGenerator)(nil), Matrix:(*v1alpha1.MatrixGenerator)(nil), Merge:(*v1alpha1.MergeGenerator)(nil)}
		1: &v1alpha1.ApplicationSet{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"", GenerateName:"", Namespace:"", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), DeletionTimestamp:<nil>, DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ClusterName:"", ManagedFields:[]v1.ManagedFieldsEntry(nil)}, Spec:v1alpha1.ApplicationSetSpec{Generators:[]v1alpha1.ApplicationSetGenerator(nil), Template:v1alpha1.ApplicationSetTemplate{ApplicationSetTemplateMeta:v1alpha1.ApplicationSetTemplateMeta{Name:"", Namespace:"", Labels:map[string]string(nil), Annotations:map[string]string(nil), Finalizers:[]string(nil)}, Spec:v1alpha1.ApplicationSpec{Source:v1alpha1.ApplicationSource{RepoURL:"", Path:"", TargetRevision:"", Helm:(*v1alpha1.ApplicationSourceHelm)(nil), Kustomize:(*v1alpha1.ApplicationSourceKustomize)(nil), Directory:(*v1alpha1.ApplicationSourceDirectory)(nil), Plugin:(*v1alpha1.ApplicationSourcePlugin)(nil), Chart:""}, Destination:v1alpha1.ApplicationDestination{Server:"", Namespace:"", Name:"", isServerInferred:false}, Project:"", SyncPolicy:(*v1alpha1.SyncPolicy)(nil), IgnoreDifferences:[]v1alpha1.ResourceIgnoreDifferences(nil), Info:[]v1alpha1.Info(nil), RevisionHistoryLimit:(*int64)(nil)}}, SyncPolicy:(*v1alpha1.ApplicationSetSyncPolicy)(nil)}, Status:v1alpha1.ApplicationSetStatus{Conditions:[]v1alpha1.ApplicationSetCondition(nil)}}

Difference found in argument 0:

--- Expected
+++ Actual
@@ -1,3 +1,17 @@
 (*v1alpha1.ApplicationSetGenerator)({
- List: (*v1alpha1.ListGenerator)(<nil>),
+ List: (*v1alpha1.ListGenerator)({
+  Elements: ([]v1.JSON) (len=1) {
+   (v1.JSON) &JSON{Raw:*[123 34 99 108 117 115 116 101 114 34 58 34 67 108 117 115 116 101 114 34 44 34 117 114 108 34 58 34 85 114 108 34 125],}
+  },
+  Template: (v1alpha1.ApplicationSetTemplate) {
+   ApplicationSetTemplateMeta: (v1alpha1.ApplicationSetTemplateMeta) {
+    Name: (string) "",
+    Namespace: (string) "",
+    Labels: (map[string]string) <nil>,
+    Annotations: (map[string]string) <nil>,
+    Finalizers: ([]string) <nil>
+   },
+   Spec: (v1alpha1.ApplicationSpec) &ApplicationSpec{Source:ApplicationSource{RepoURL:,Path:,TargetRevision:,Helm:nil,Kustomize:nil,Directory:nil,Plugin:nil,Chart:,},Destination:ApplicationDestination{Server:,Namespace:,Name:,},Project:,SyncPolicy:nil,IgnoreDifferences:[]ResourceIgnoreDifferences{},Info:[]Info{},RevisionHistoryLimit:nil,}
+  }
+ }),
  Clusters: (*v1alpha1.ClusterGenerator)(<nil>),

Diff: 0: FAIL:  (*v1alpha1.ApplicationSetGenerator=&{0xc000148b00 <nil> 0xc000703860 <nil> <nil> <nil> <nil> <nil>}) != (*v1alpha1.ApplicationSetGenerator=&{<nil> <nil> 0xc000703040 <nil> <nil> <nil> <nil> <nil>})
	1: PASS:  (*v1alpha1.ApplicationSet=&{{ } {      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []} {[] {{  map[] map[] []} {{   nil nil nil nil } {   false}  nil [] [] <nil>}} <nil>} {[]}}) == (*v1alpha1.ApplicationSet=&{{ } {      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []} {[] {{  map[] map[] []} {{   nil nil nil nil } {   false}  nil [] [] <nil>}} <nil>} {[]}})

Process finished with the exit code 1
```

KojoRising commented:
Ok, found the issue. It's because the interpolatedGenerator creates a copy of the original generator. And so because it's a new object, the mocked call isn't actually ever triggered (because the object is different).


Here's the failing test-case for example: 
<img width="1309" alt="image" src="https://user-images.githubusercontent.com/29113963/167315881-b6c8dbca-7ac5-407a-ab7d-8b7b75eb58bc.png">



And here's the difference when we just return the requestedGenerator that was passed in:
 
<img width="1251" alt="image" src="https://user-images.githubusercontent.com/29113963/167315955-ea493600-4712-47d9-b7c1-3bb62639a2a2.png">

KojoRising commented:
@crenshaw-dev  @jgwest  - I had seen you folks had done a lot of work on the original matrix-test found here: https://github.com/argoproj/applicationset/commits/master/pkg/generators/matrix_test.go


Can y'all take a look & provide some feedback/direction to go in? I cannot get that last test case fixed for whatever reason (been trying for a couple of days now); I believe it has something to do with the GoLang mocked call not triggering correctly, but I'm unable to figure it out.

jannis-a commented:
@KojoRising @ashutosh16 @crenshaw-dev @jgwest 

I might have an idea: how about adding additional checks to the nested generators before actually trying to generate parameters and fail earlier? The mock is then not running into the issue having to deal with the `interpolatedGenerator`. From how I've understood the code, this could do the trick and makes the unit tests pass.

This could be maybe solved a bit more elegant but shows the idea:
```diff
diff --git a/applicationset/generators/generator_spec_processor.go b/applicationset/generators/generator_spec_processor.go
index 9b04f3502..43a4efaac 100644
--- a/applicationset/generators/generator_spec_processor.go
+++ b/applicationset/generators/generator_spec_processor.go
@@ -66,7 +66,7 @@ func Transform(requestedGenerator argoprojiov1alpha1.ApplicationSetGenerator, al
        return res, firstError
 }

-func GetRelevantGenerators(requestedGenerator *argoprojiov1alpha1.ApplicationSetGenerator, generators map[string]Generator) []Generator {
+func getRelevantGenerators(requestedGenerator interface{}, generators map[string]Generator) []Generator {
        var res []Generator

        v := reflect.Indirect(reflect.ValueOf(requestedGenerator))
@@ -84,6 +84,14 @@ func GetRelevantGenerators(requestedGenerator *argoprojiov1alpha1.ApplicationSet
        return res
 }

+func getRelevantNestedGenerators(requestedGenerator argoprojiov1alpha1.ApplicationSetNestedGenerator, generators map[string]Generator) []Generator {
+       return getRelevantGenerators(requestedGenerator, generators)
+}
+
+func GetRelevantGenerators(requestedGenerator *argoprojiov1alpha1.ApplicationSetGenerator, generators map[string]Generator) []Generator {
+       return getRelevantGenerators(requestedGenerator, generators)
+}
+
 func mergeGeneratorTemplate(g Generator, requestedGenerator *argoprojiov1alpha1.ApplicationSetGenerator, applicationSetTemplate argoprojiov1alpha1.ApplicationSetTemplate) (argoprojiov1alpha1.ApplicationSetTemplate, error) {

        // Make a copy of the value from `GetTemplate()` before merge, rather than copying directly into
diff --git a/applicationset/generators/matrix.go b/applicationset/generators/matrix.go
index 1c9d9a5be..6fb1403af 100644
--- a/applicationset/generators/matrix.go
+++ b/applicationset/generators/matrix.go
@@ -42,6 +42,14 @@ func (m *MatrixGenerator) GenerateParams(appSetGenerator *argoprojiov1alpha1.App
                return nil, ErrMoreThanTwoGenerators
        }

+       if len(getRelevantNestedGenerators(appSetGenerator.Matrix.Generators[0], m.supportedGenerators)) > 1 {
+               return nil, ErrMoreThenOneInnerGenerators
+       }
+
+       if len(getRelevantNestedGenerators(appSetGenerator.Matrix.Generators[1], m.supportedGenerators)) > 1 {
+               return nil, ErrMoreThenOneInnerGenerators
+       }
+
        res := []map[string]string{}

        g0, err := m.getParams(appSetGenerator.Matrix.Generators[0], appSet, map[string]string{})
```

```
/usr/local/opt/go/libexec/bin/go tool test2json -t /private/var/folders/jm/qcpv3yp11fzd4g5wcwwcnqs5dh1_66/T/GoLand/___1TestMatrixGenerate_in_github_com_argoproj_argo_cd_v2_applicationset_generators.test -test.v -test.paniconexit0 -test.run ^\QTestMatrixGenerate\E$
=== RUN   TestMatrixGenerate
=== RUN   TestMatrixGenerate/happy_flow_-_generate_params
=== RUN   TestMatrixGenerate/happy_flow_-_generate_params_from_two_lists
=== RUN   TestMatrixGenerate/returns_error_if_there_is_less_than_two_base_generators
=== RUN   TestMatrixGenerate/returns_error_if_there_is_more_than_two_base_generators
=== RUN   TestMatrixGenerate/returns_error_if_there_is_more_than_one_inner_generator_in_the_first_base_generator
=== RUN   TestMatrixGenerate/returns_error_if_there_is_more_than_one_inner_generator_in_the_second_base_generator
--- PASS: TestMatrixGenerate (0.00s)
    --- PASS: TestMatrixGenerate/happy_flow_-_generate_params (0.00s)
    --- PASS: TestMatrixGenerate/happy_flow_-_generate_params_from_two_lists (0.00s)
    --- PASS: TestMatrixGenerate/returns_error_if_there_is_less_than_two_base_generators (0.00s)
    --- PASS: TestMatrixGenerate/returns_error_if_there_is_more_than_two_base_generators (0.00s)
    --- PASS: TestMatrixGenerate/returns_error_if_there_is_more_than_one_inner_generator_in_the_first_base_generator (0.00s)
    --- PASS: TestMatrixGenerate/returns_error_if_there_is_more_than_one_inner_generator_in_the_second_base_generator (0.00s)
PASS

Process finished with the exit code 0
````

rumstead commented:
Another approach which may be more elegant? Since we are already using the `testify` framework.

```diff
diff --git a/applicationset/generators/matrix_test.go b/applicationset/generators/matrix_test.go
index c5c7fce24..0b02ff483 100644
--- a/applicationset/generators/matrix_test.go
+++ b/applicationset/generators/matrix_test.go
@@ -127,7 +127,7 @@ func TestMatrixGenerate(t *testing.T) {
                testCaseCopy := testCase // Since tests may run in parallel
 
                t.Run(testCaseCopy.name, func(t *testing.T) {
-                       mock := &generatorMock{}
+                       generatorMock := &generatorMock{}
                        appSet := &argoprojiov1alpha1.ApplicationSet{}
 
                        for _, g := range testCaseCopy.baseGenerators {
@@ -136,7 +136,7 @@ func TestMatrixGenerate(t *testing.T) {
                                        Git:  g.Git,
                                        List: g.List,
                                }
-                               mock.On("GenerateParams", &gitGeneratorSpec, appSet).Return([]map[string]string{
+                               generatorMock.On("GenerateParams", mock.AnythingOfType("*v1alpha1.ApplicationSetGenerator"), appSet).Return([]map[string]string{
                                        {
                                                "path":                    "app1",
                                                "path.basename":           "app1",
@@ -149,13 +149,13 @@ func TestMatrixGenerate(t *testing.T) {
                                        },
                                }, nil)
 
-                               mock.On("GetTemplate", &gitGeneratorSpec).
+                               generatorMock.On("GetTemplate", &gitGeneratorSpec).
                                        Return(&argoprojiov1alpha1.ApplicationSetTemplate{})
                        }
 
                        var matrixGenerator = NewMatrixGenerator(
                                map[string]Generator{
-                                       "Git":  mock,
+                                       "Git":  generatorMock,
                                        "List": &ListGenerator{},
                                },
                        )
```
```
=== RUN   TestMatrixGenerate
=== RUN   TestMatrixGenerate/happy_flow_-_generate_params
=== RUN   TestMatrixGenerate/happy_flow_-_generate_params_from_two_lists
=== RUN   TestMatrixGenerate/returns_error_if_there_is_less_than_two_base_generators
=== RUN   TestMatrixGenerate/returns_error_if_there_is_more_than_two_base_generators
=== RUN   TestMatrixGenerate/returns_error_if_there_is_more_than_one_inner_generator_in_the_first_base_generator
=== RUN   TestMatrixGenerate/returns_error_if_there_is_more_than_one_inner_generator_in_the_second_base_generator
--- PASS: TestMatrixGenerate (0.00s)
    --- PASS: TestMatrixGenerate/happy_flow_-_generate_params (0.00s)
    --- PASS: TestMatrixGenerate/happy_flow_-_generate_params_from_two_lists (0.00s)
    --- PASS: TestMatrixGenerate/returns_error_if_there_is_less_than_two_base_generators (0.00s)
    --- PASS: TestMatrixGenerate/returns_error_if_there_is_more_than_two_base_generators (0.00s)
    --- PASS: TestMatrixGenerate/returns_error_if_there_is_more_than_one_inner_generator_in_the_first_base_generator (0.00s)
    --- PASS: TestMatrixGenerate/returns_error_if_there_is_more_than_one_inner_generator_in_the_second_base_generator (0.00s)
PASS
```

KojoRising commented:
@crenshaw-dev, @Lobstrosity, @rishabh625

If y'all want to give a final look through, that'd be great! 

But LGTM from my end, thanks @rumstead and @jannis-a for helping getting this over the finish line!

rishabh625 commented on docs/operator-manual/applicationset/Generators-Matrix.md:
```diff
@@ -104,18 +104,80 @@ Finally, the Matrix generator will combine both sets of outputs, and produce:
 
 1. The Matrix generator currently only supports combining the outputs of only two child generators (eg does not support generating combinations for 3 or more).
-1. You should specify only a single generator per array entry, eg this is not valid:
+2. You should specify only a single generator per array entry, eg this is not valid:
```
please don't update the numbers

KojoRising replied:
Should I change them all back to "1." ? I thought that was an error?

rishabh625 replied:
Yeah, it's like that, check main Readme.md it will get rendered in ordered way

KojoRising replied:
Sounds good, will change it

KojoRising replied:
Ok, should be done

rishabh625 reviewed (approved):
LGTM , other than comment on doc.
 @crenshaw-dev: could help us to get it through finish line

crenshaw-dev commented:
@rishabh625 reviewing!

crenshaw-dev commented on applicationset/generators/generator_spec_processor.go:
```diff
@@ -81,3 +94,31 @@ func mergeGeneratorTemplate(g Generator, requestedGenerator *argoprojiov1alpha1.
+	tmplBytes, err := json.Marshal(interpolatedGenerator)
+	if err != nil {
+		log.WithError(err).WithField("requestedGenerator", interpolatedGenerator).Error("error marshalling requested generator for interpolation")
+		return *interpolatedGenerator, err
```
Can we just return `nil, err`? I don't think we end up using the interpolatedGenerator's value if there's an error (we hit the `continue` instead and skip interpolation).

KojoRising replied:
That was actually my original plan (to use pointers). What we noticed is that in a scenario where you have say, two git generator paths matrixed w/ a cluster generator w/ the placeholder parameter, on the first pass when using a pointer it replaced the placeholder with the actual value altogether. So a label like this ` kubernetes.io/environment: '{{path.basename}}' ` was replaced by say `dev` on the 1st pass. So on the second pass, it would try to interpolate ` kubernetes.io/environment: dev `, but since the placeholder was already replaced, it would skip it altogether.

crenshaw-dev commented on applicationset/generators/generator_spec_processor.go:
```diff
@@ -81,3 +94,31 @@ func mergeGeneratorTemplate(g Generator, requestedGenerator *argoprojiov1alpha1.
+	replacedTmplStr, err := render.Replace(fstTmpl, params, true)
+	if err != nil {
+		log.WithError(err).WithField("interpolatedGeneratorString", replacedTmplStr).Error("error interpolating generator with other generator's parameter")
+		return *interpolatedGenerator, err
```
Ditto.

KojoRising replied:
Resolved!  (as per earlier conversation)

crenshaw-dev commented on applicationset/generators/generator_spec_processor.go:
```diff
@@ -81,3 +94,31 @@ func mergeGeneratorTemplate(g Generator, requestedGenerator *argoprojiov1alpha1.
+	err = json.Unmarshal([]byte(replacedTmplStr), interpolatedGenerator)
+	if err != nil {
+		log.WithError(err).WithField("requestedGenerator", interpolatedGenerator).Error("error unmarshalling requested generator for interpolation")
+		return *interpolatedGenerator, err
```
Ditto.

KojoRising replied:
Resolved!  (as per earlier conversation)

crenshaw-dev commented on applicationset/generators/matrix.go:
```diff
@@ -44,16 +44,15 @@ func (m *MatrixGenerator) GenerateParams(appSetGenerator *argoprojiov1alpha1.App
 	res := []map[string]string{}
 
-	g0, err := m.getParams(appSetGenerator.Matrix.Generators[0], appSet)
+	g0, err := m.getParams(appSetGenerator.Matrix.Generators[0], appSet, map[string]string{})
```
nit: I think you can just pass `nil` for the map.

KojoRising replied:
Resolved!

crenshaw-dev reviewed (changes requested):
This looks awesome! I added a few nitpicks.

The only substantial change I'd like to see is a full test in matrix_test.go, similar to the on already in `TestMatrixGenerate`. I think there's sufficient mock logic available to test the exact same appset as what you added to the matrix generator docs.

KojoRising commented on applicationset/generators/generator_spec_processor.go:
```diff
@@ -81,3 +94,31 @@ func mergeGeneratorTemplate(g Generator, requestedGenerator *argoprojiov1alpha1.
+
+// Currently for Matrix Generator. Allows interpolating the matrix's 2nd child generator with values from the 1st child generator
+// "params" parameter is an array, where each index corresponds to a generator. Each index contains a map w/ that generator's parameters.
+// Since the Matrix generator currently only allows 2 child generators, effectively this params array will always be size 1 (containing the 1st child generator's params)
```
Resolved!

crenshaw-dev commented:
@KojoRising you're an absolute legend. I'm going to give it one last pass, and unless I notice something completely wild, will approve and set to auto-merge.  :-)

crenshaw-dev commented:
@KojoRising can you fix the DCO check?

crenshaw-dev commented:
@KojoRising fixing the DCO check is a pain. Looks like just this one commit message needs to be fixed:

> Commit sha: [8427c47](https://github.com/argoproj/argo-cd/pull/9080/commits/8427c4765476b05dee288ff0b52e0116b6fe47b6), Author: Jay P Kulkarni, Committer: GitHub; Expected "Jay P Kulkarni [jkulkarni@ucla.edu](mailto:jkulkarni@ucla.edu)", but got "jkulkarn [jay.p.kulkarni@blackrock.com](mailto:jay.p.kulkarni@blackrock.com)".

KojoRising commented:
Thanks @crenshaw-dev - I was about to ask you for some advice there. I'll try fixing that commit message rn. I was trying these steps from the DCO docs:

1. Ensure you have a local copy of your branch by [checking out the pull request locally via command line](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/checking-out-pull-requests-locally).
2. In your local branch, run: git rebase HEAD~219 --signoff
3. Force push your changes to overwrite the branch: git push --force-with-lease origin feat/add-intra-generator-param-support

KojoRising commented:
@crenshaw-dev any advice here? I was trying to find some resources to help, but I think it just created new commits w/ the changed Author, rather than fixing that commit (I still see DCO identifying that particular commit as an issue). 
https://www.git-tower.com/learn/git/faq/change-author-name-email
https://rushlow.dev/blog/oops-i-forgot-to-sign-my-commit-from-last-monday


![image](https://user-images.githubusercontent.com/29113963/172211318-b33d2fbc-f18d-4cb9-92d3-c0103eafcafb.png)

crenshaw-dev commented:
I think it's gonna be tedious... basically `git rebase -i 2c69828` and change the `pick` to `edit` for the two problem commits. https://stackoverflow.com/a/1186549/684776

I you wanna hop on a call, feel free to find me on CNCF Slack!  :-)

crenshaw-dev reviewed (approved):
LGTM! Thanks so much all, especially @KojoRising!

KojoRising commented:
Haha, thanks so much @crenshaw-dev!! Appreciate all the reviewing, it's been a long, but fun PR. 

Thanks again @Lobstrosity, @rishabh625, @jannis-a, @rumstead for helping me get through this!!!

michael-bowen-sc commented:
> LGTM! Thanks so much all, especially @KojoRising!

@KojoRising  @rumstead  amazing stuff gents!

gaeljw commented:
Thanks everyone for the work. In which version of ArgoCD will this available? 2.4.1?

crenshaw-dev commented:
@gaeljw since it's a new feature rather than a bugfix, I plan to let it wait until 2.5.0, which should be released in August.

gaeljw commented:
Sure, good to know @crenshaw-dev :ok_hand:

olegeech-me commented:
Sorry for a dumb question folks, is there a dev image pushed somewhere with this feature inside? @KojoRising @crenshaw-dev 
Will `k apply -f https://raw.githubusercontent.com/argoproj/applicationset/master/manifests/install.yaml` work for me?
I'd like to get use of this feature before it's released in 2.5.

rishabh625 commented:
applying above manifests won't work @olegeech-me , but I think , but applying manifests from this repo should work
kubectl apply -f manifests/install.yaml

KojoRising commented:
Also @olegeech-me if you wanna try via kustomize (that's how I currently do it): You can just pin it to the latest commit and do a `kustomize build [KUSTOMIZE-DIRECTOY]` and it should work fine (correct me if I'm wrong @rishabh625)

```yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
namespace: argocd
resources:
- https://github.com/argoproj/argo-cd/manifests/ha/cluster-install?ref=2b5371681f611ca05c1884000bd236a38e02f167
```

Or non-pinned latest
```yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
namespace: argocd
resources:
- https://github.com/argoproj/argo-cd/manifests/ha/cluster-install
```

jeromepin commented:
Is this going to work for merge generators too ?

crenshaw-dev commented:
@jeromepin I believe this work was limited to only matrix generators, but I could be remembering wrong.
